### ISSUE TITLE
Allow to fail script on after_deploy

### ISSUE THREAD

**Issue by drogus** [travis-ci/travis-ci#1574]
Return value of after_deploy commands does not have any effect on the result of the script. However, it is a valid scenario to check if deploy succeeded (for example by sending a request to a deployed app) and fail the build if this is not the case.

This is not likely going to be a default, but we could allow to configure this somehow in .travis.yml

**Comment by aripollak** [comment]
Are you sure this is a Travis problem and not [this](https://github.com/heroku/heroku/issues/186) Heroku Toolbelt one?

**Comment by ryanbrainard** [comment]
@aripollak Yes, this is happening for non-Heroku post deploy scripts.

**Comment by ericminio** [comment]
@BanzaiMan Hi, 
Will you share the intentions behind allowing a after_deploy script without taking into account its exit code?

**Comment by ryanbrainard** [comment]
@joshk, any insight we can get here on @ericminio's question or this issue as a whole? It really feels like a bug with Travis and there's obviously been a lot of people affected by it from looking at this issue. If you don't want to change existing behavior, I'd be happy with some kind of `failed_post_deploy_script_fails_build: true` in my config if need be.

**Comment by localnerve** [comment]
@BanzaiMan Please consider my use case: I deploy to a staging location to run further tests using other services. These tests are just as important as the unit tests, but rely on the deployment.
after_deploy, I run functional tests at saucelabs and performance budget tests using wpt. I would like to add even more tests against the staged app.
The Travis deployment feature opened the door to this level of test automation. The ability to fail the build from after_deploy is a requirement to make full use of this.

**Comment by i8ramin** [comment]
+1
would love to know if the heroku deploy was successful and then trigger some automated functional tests using saucelabs as well. Any updates on this?

**Comment by rfarine** [comment]
This worked for us (using nightwatch to run our tests)

```
after_deploy:
  - "if [[ \"$TRAVIS_BRANCH\" == \" branch name here \" ]]; then bundle exec nightwatch -e saucelabs || travis_terminate 1; else echo \"Skipping nightwatch js end to end testing\"; fi;"
```

**Comment by rolebi** [comment]
@rfarine never heard of this one, is this documented anywhere ? Nice trick BTW  :)

**Comment by localnerve** [comment]
@BanzaiMan / Travis-ci,

1. Is `travis_terminate` always available in Travis environments?
2. Is `travis_terminate` the 'blessed' procedure for failing in `after_deploy`?

Thanks in advance for the info.

**Comment by rfarine** [comment]
@rolebi No, actually! We (here at WeWork) emailed the team at Travis CI and they recommended we take that route. :) and thanks!

**Comment by BanzaiMan** [comment]
`travis_terminate` is currently undocumented, but it is unlikely to disappear.

**Comment by dtothefp** [comment]
@BanzaiMan another use case is that we use the `deploy` command to deploy static assets to S3 to path on our dev server that matches the branch name. We then run end to end tests with selenium on that path. This is nice for us because although failing tests don't block the build deployment, we can test all our branches on before merging to master. Unfortunately, exiting the process with a code other than `0` seems to have no effect and it would be nice to see in the github interface for the branch/pr a failing build if tests do not pass.

**Comment by joshk** [comment]
I this we should add this to the deployment provider. It could be an option, eg. `fail_build: true`

@BanzaiMan what are your thoughts on this?

**Comment by BanzaiMan** [comment]
How about setting a published environment variable (`$TRAVIS_DEPLOY_RESULT` might be a good name), which can then be used by `after_deploy` to act accordingly?

Where would `fail_build: true` go?

**Comment by joshk** [comment]
It would go in the deployment config.

Totally :+1: on the env var too.

**Comment by nfwlpw-mit** [comment]
Not just capturing the exit status but probably other results as well, eg I'm using AWS Codedeploy, after deploy is done all I get is: 

Triggered deployment "d-8BYLL3NDD"

I need to be able to capture the deployment-id and pass it onto a post deployment script.

**Comment by rolebi** [comment]
@nfwlpw-mit if it can help, we use some bash scripts to accomplish that: https://gist.github.com/rolebi/2986d0899a526fd6cfe6 works pretty well.

**Comment by nfwlpw-mit** [comment]
Thank you Ronan, I'll take a look.

**Comment by pkyeck** [comment]
what is the status on this? really looking forward to any option that would allow checking if a deploy succeded.

**Comment by kelseyq** [comment]
seconded--we also would like to use after_deploy or after_script for functional testing on a deployed app

**Comment by hassankhan** [comment]
Would love to see this added. Could it be something as simple as checking the build command's result code?

**Comment by matthewcummings** [comment]
It's 2017 and I would like this too.  :)

**Comment by tdmalone** [comment]
@rolebi The gist you linked to above is no longer available. Do you still have those examples anywhere?

**Comment by rolebi** [comment]
@tdmalone restored : )

**Comment by tdmalone** [comment]
^ Not resolved, would still love to have the ability for after_deploy to fail the build

**Comment by kanekotic** [comment]
is 2018 and this is still not solved and seems a very normal use case. When you are deploying, if it doesn't work, currently you don't get any feedback. 
My case is I expect my packages to have been deployed to NPM and they weren't because Travis failed silently. Sincerely I don't see any case where anybody would like their deployment step to fail silently.
Shouldn't the default behavior be the opposite?

**Comment by aripalo** [comment]
Or if changing the behavior of `after_deploy` is unwanted, then maybe provide us a new lifecycle hook, maybe `validate_deploy`...?

**Comment by aripalo** [comment]
Something like `validate_deploy` would make quite a lot of sense, as you could provide support for building rollback and or blue/green features.

Like:

```
validate_deploy:
  script: "echo 'validate'"
  on_success:
  - "echo 'here one could configure steps e.g. for blue/green change'" 
  on_fail: 
  - "echo 'here one could configure rollback steps'" 
```

**Comment by Elijas** [comment]
WHY is this not a thing after 5 years

**Comment by tdmalone** [comment]
@Elijas If you're just getting started.. I'd suggest looking at Circle CI instead. Seem to integrate/improve stuff a lot faster than Travis is these days (I've been meaning to move but just haven't had a chance to learn it yet)

**Comment by Elijas** [comment]
Oh, I've missed the comments above about the `travis_terminate`, works perfectly.

travis.yml:
```
after_deploy:
  - if ! TEST_SCRIPT_HERE; then travis_terminate 1; fi
```

> @Elijas If you're just getting started.. I'd suggest looking at Circle CI instead. Seem to integrate/improve stuff a lot faster than Travis is these days (I've been meaning to move but just haven't had a chance to learn it yet)

Thank you! Although Travis seems to be working fine for now, I might look into it later 👍
